mesh/nurbs.cpp

```diff
@@ -85,18 +92,138 @@ KnotVector *KnotVector::DegreeElevate(int t) const
    return newkv;
 }
 
-void KnotVector::UniformRefinement(Vector &newknots) const
+void KnotVector::UniformRefinement(Vector &newknots, int rf) const
 {
-   newknots.SetSize(NumOfElements);
+   MFEM_VERIFY(rf > 1, "Refinement factor must be at least 2.");
+
+   const double h = 1.0 / ((double) rf);
+
+   newknots.SetSize(NumOfElements * (rf - 1));
    int j = 0;
    for (int i = 0; i < knot.Size()-1; i++)
    {
       if (knot(i) != knot(i+1))
       {
-         newknots(j) = 0.5*(knot(i) + knot(i+1));
-         j++;
+         for (int m = 1; m < rf; ++m)
+         {
+            newknots(j) = m * h * (knot(i) + knot(i+1));
+            j++;
+         }
+      }
+   }
+}
+
+int KnotVector::GetCoarseningFactor() const
+{
+   if (spacing)
+   {
+      if (spacing->Nested())
+      {
+         return 1;
+      }
+      else
+      {
+         return spacing->Size();   // Coarsen only if non-nested
+      }
+   }
+   else
+   {
+      return 1;
+   }
+}
+
+void KnotVector::GetFineKnots(const int cf, Vector & fine)
```

Comment:
Interesting suggestion, but I'm not sure about the definition of "small" here. Typically, a 2D or 3D NURBS patch is defined by a tensor product of knot vectors, so the number of elements in 2D would be n1 * n2 with 2 knot vectors of size on the order of n1 and n2. This means the number of knots is relatively small compared to the overall computation, but it is not O(1). The number of fine knots here is O(n) for the knot vector length n. In the case of 1D applications, which are not really a performance concern, that O(n) cost is more significant. So I think it is better to pass by reference, although passing by value should not be a bottleneck.

Reply:
Ah I wasn't clear enough, I meant to declare the `Vector fine` inside the method, and return it, rather than have to perform the declaration outside and use an out parameter. Out parameters can be warranted if repeated initialization can be avoided/is costly, but the call site for this method the `Vector` is being declared inside the loop already. Basically, I'm arguing that using an out parameter is a premature optimization.

Reply:
I implemented your suggestion.